### ISSUE TITLE
Presigned PUT - SignatureDoesNotMatch HTTP 403

### ISSUE THREAD

**Issue by adam1010** [aws/aws-sdk-php#1344]
I'm using the latest V3 version (SDK v3.32.3)

```
$s3 = \Aws\S3\S3Client::factory(...)
$cmd = $s3->getCommand('PutObject', 
     array( 'Bucket'=>'test-bucket', 
               'Key'=>'test.jpg', 
               'ContentType'=>'binary/octet-stream')
);    
$request = $s3->createPresignedRequest($cmd, "+6 hours");
$upload_url =  (string) $request->getUri();
```

// Test Via:
`curl -v -H "Content-Type: binary/octet-stream" -T test.jpg "$upload_url"`

// Output:

```
HTTP/1.1 403 Forbidden

<Error>
   <Code>SignatureDoesNotMatch</Code>
  <Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
  <AWSAccessKeyId>xxxx</AWSAccessKeyId>
  <StringToSign>xxx</StringToSign>
   ....
</Error>
```

**Comment by kstich** [comment]
Hi @adam1010 , I'm not able to reproduce your issue while using the same SDK version as you. Are you possibly doing any other manipulation to the URL before using it for the cURL call? Also, please make sure that you're copying all of the characters of the URL when performing the cURL, leaving a single character off the signature parameter (usually the last in the string) will cause the error you're seeing.

**Comment by adam1010** [comment]
Thanks @kstich -- No, I'm not doing any manipulation to the URL. I've tried it a few dozen times so I don't think it's an incomplete url problem.

I'm using PHP 5.6.25

Here's the full code for how I initialize the S3 object:
`$s3 = \Aws\S3\S3Client::factory(array('version'=>'latest','region'=>'us-east-1','http'=>'/cert.pem', 'credentials'=> ['key'=>'xxx','secret'=>'xxx']))`

Here's an example of the URL it spits out:
`https://xyz.s3.amazonaws.com/temp.jpg?X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=XXXJHZIS4IL7K46OOAA%2F20170807%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170807T204004Z&X-Amz-SignedHeaders=host&X-Amz-Expires=43200&X-Amz-Signature=7e916541d17ef5ba9a6380fe7146e6ab61922f4c98d8d00bf4ca5fc3816a072f1`

Do you have any other ideas or theories I could test?

**Comment by adam1010** [comment]
I fired up the signing code in a debugger and compared the output of curl's error response (in the <CanonicalRequest> section) to what the SDK is signing (the $context['creq'] variable on line 98 of src/Signature/SignatureV4.php)

I noticed they are NOT identical.  The SDK canonical request did not urlencode the ampersands  so they appear as &  instead of &amp;   (whereas the error output from AWS via curl has them encoded as &amp;) -- Also, the X-Amz-Date  and X-Amz-Expires  values to not match.   Not sure if either of these are relevant but I thought I'd mention it in case there is something different about my PHP configuration causing the issue.
